File: hooks/scratch.py
#!/usr/bin/env python3
"""Keep a project's scratch directory from becoming a graveyard.

An agent writes probes, captures, half-finished drafts and one-off scripts. Sent
to the system temporary directory they are invisible to whoever is reviewing the
work; left at the root of the project's own `.tmp/` they pile up, and nothing
ever tells one throwaway apart from a draft that still matters.

So a session writes its throwaway under a directory named for itself, and that
directory goes when the session ends. Anything worth keeping is written
somewhere else under `.tmp/`, where nothing here can reach it.

    scratch.py end     remove this session's own directory
    scratch.py start   reap directories abandoned by sessions that crashed,
                       and report what the memory directory has let slip

**Only a directory named for a session is ever removed.** A name this cannot
parse as a session identifier is somebody's deliberate keep, and is left alone
whatever its age.
"""

import json
import os
import re
import shutil
import subprocess
import sys
import time
from pathlib import Path

# A session identifier as the client writes it. Matching this shape is the whole
# safety of the sweep: a directory named anything else is never removed.
SESSION = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")

# How long a directory whose session never ended stays. Long enough that a
# session resumed the next morning still finds its own work.
ORPHAN_DAYS = 7

# Past this a carry-forward has stopped being a handover and become a knowledge
# store, which goes stale and then gets believed.
CARRY_LIMIT = 20 * 1024


def repo_root():
    """The main checkout, even for a session driven inside a worktree."""
    base = Path(os.environ.get("CLAUDE_PROJECT_DIR") or Path.cwd())
    done = subprocess.run(
        ["git", "rev-parse", "--path-format=absolute", "--git-common-dir"],
        capture_output=True, text=True, check=False, cwd=base,
    )
    if done.returncode == 0 and done.stdout.strip():
        return Path(done.stdout.strip()).parent

    return base


def scratch_dir():
    return repo_root() / ".tmp"


def ours():
    return scratch_dir() / "claude"


def enabled():
    """A project opts in by carrying a scratch directory. Nothing is created."""
    return scratch_dir().is_dir()


def session_of(payload):
    return str(payload.get("session_id") or "").strip()


def remove(directory):
    """Drop one directory, reporting nothing a caller cannot act on."""
    try:
        shutil.rmtree(directory)
        return True
    except OSError:
        return False


def command_end(payload):
    """Remove the directory this session wrote its throwaway into."""
    session = session_of(payload)
    if not SESSION.match(session):
        return 0

    held = ours() / session
    if held.is_dir() and remove(held):
        print(f"swept {held}")

    return 0


def reap():
    """Every session directory old enough that nobody is still writing it."""
    if not ours().is_dir():
        return []

    stale = time.time() - ORPHAN_DAYS * 86400
    gone = []
    for held in sorted(ours().iterdir()):
        if not held.is_dir() or not SESSION.match(held.name):
            continue
        if held.stat().st_mtime < stale and remove(held):
            gone.append(held.name)

    return gone


def slipped():
    """What the memory directory has let slip, in the words a reader acts on."""
    memory = repo_root() / ".memory"
    if not memory.is_dir():
        return []

    found = []
    carry = memory / "carryforward"
    if carry.is_dir():
        for held in sorted(carry.glob("*-memory.md")):
            size = held.stat().st_size
            if size > CARRY_LIMIT:
                found.append(f"{held.name} is {size // 1024} KB, past the {CARRY_LIMIT // 1024} KB "
                             "a carry-forward holds. Anything the repository can rebuild goes.")

    index = memory / "MEMORY.md"
    if index.is_file():
        for line in index.read_text(encoding="utf-8", errors="replace").splitlines():
            for target in re.findall(r"\]\(([^)]+)\)", line):
                if target.startswith(("http://", "https://")):
                    continue
                if not (memory / target).exists():
                    found.append(f"MEMORY.md points at {target}, which is not there.")

    return found


def command_start(payload):
    """Reap what crashed sessions left, and say what memory has let slip."""
    said = []
    gone = reap()
    if gone:
        said.append(f"Scratch: reaped {len(gone)} session directory(s) nobody came back to.")
    said.extend(slipped())
    if said:
        print("\n".join(said))

    return 0


MODES = {"end": command_end, "start": command_start}


def main():
    if len(sys.argv) < 2 or sys.argv[1] not in MODES:
        print(f"usage: {sys.argv[0]} {'|'.join(MODES)}", file=sys.stderr)
        return 1

    if not enabled():
        return 0

    try:
        payload = json.load(sys.stdin)
    except ValueError:
        payload = {}
    if not isinstance(payload, dict):
        payload = {}

    return MODES[sys.argv[1]](payload)


if __name__ == "__main__":
    sys.exit(main())

File: source/hooks/internal/scratch/scratch.go
// Package scratch keeps a project's scratch directory from becoming a
// graveyard.
//
// A session writes its throwaway under `.tmp/claude/<session-id>/`, and that
// directory goes when the session ends. A directory abandoned by a session that
// crashed is reaped at a later session start once it is a week old.
//
// Only a directory named for a session is ever removed. A name that does not
// parse as a session identifier is somebody's deliberate keep, and survives
// whatever its age.
package scratch

import (
	"context"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/fza/agentic-helpers/source/hooks/internal/hookio"
)

// Matching this shape is the whole safety of the sweep.
var sessionPattern = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)

var memoryLink = regexp.MustCompile(`\]\(([^)]+)\)`)

var ErrUsage = errors.New("usage: agentic-scratch end|start")

// Long enough that a session resumed the next morning still finds its work.
const orphanAge = 7 * 24 * time.Hour

// Past this a carry-forward has stopped being a handover and become a
// knowledge store, which goes stale and then gets believed.
const carryLimit = 20 * 1024

func isSession(name string) bool {
	return sessionPattern.MatchString(name)
}

// Main runs one hook invocation against the checkout at root. A project opts
// in by carrying `.tmp/`; without it nothing is read, created or said.
func Main(ctx context.Context, args []string, root string, streams hookio.Streams) int {
	if len(args) != 1 || (args[0] != "end" && args[0] != "start") {
		_, _ = fmt.Fprintln(streams.Err, ErrUsage)

		return 1
	}

	if !isDir(filepath.Join(root, ".tmp")) {
		return 0
	}

	// A payload the client garbled names no session, and removes nothing.
	payload, err := hookio.ReadPayload(streams.In)
	if err != nil {
		payload = hookio.Payload{}
	}

	if args[0] == "end" {
		err = end(ctx, root, payload.SessionID, streams.Out)
	} else {
		err = start(ctx, root, streams.Out)
	}

	if err != nil {
		_, _ = fmt.Fprintln(streams.Err, err)

		return 1
	}

	return 0
}

func sessionsDir(root string) string {
	return filepath.Join(root, ".tmp", "claude")
}

func end(ctx context.Context, root string, session string, out io.Writer) error {
	if !isSession(session) {
		return nil
	}

	held := filepath.Join(sessionsDir(root), session)
	if !isDir(held) || !remove(ctx, held) {
		return nil
	}

	_, err := fmt.Fprintf(out, "swept %s\n", held)
	if err != nil {
		return fmt.Errorf("reporting the sweep: %w", err)
	}

	return nil
}

func start(ctx context.Context, root string, out io.Writer) error {
	var said []string

	gone := reap(ctx, root, time.Now())
	if gone > 0 {
		said = append(said, fmt.Sprintf("Scratch: reaped %d session directory(s) nobody came back to.", gone))
	}

	said = append(said, slipped(filepath.Join(root, ".memory"))...)
	if len(said) == 0 {
		return nil
	}

	_, err := fmt.Fprintln(out, strings.Join(said, "\n"))
	if err != nil {
		return fmt.Errorf("reporting the start sweep: %w", err)
	}

	return nil
}

func reap(ctx context.Context, root string, now time.Time) int {
	entries, err := os.ReadDir(sessionsDir(root))
	if err != nil {
		return 0
	}

	gone := 0
	stale := now.Add(-orphanAge)

	for _, entry := range entries {
		if ctx.Err() != nil {
			break
		}

		if !entry.IsDir() || !isSession(entry.Name()) {
			continue
		}

		info, err := entry.Info()
		if err != nil || !info.ModTime().Before(stale) {
			continue
		}

		if remove(ctx, filepath.Join(sessionsDir(root), entry.Name())) {
			gone++
		}
	}

	return gone
}

// slipped reports what the memory directory has let slip, in the words a
// reader acts on.
func slipped(memory string) []string {
	if !isDir(memory) {
		return nil
	}

	var found []string

	carries, _ := filepath.Glob(filepath.Join(memory, "carryforward", "*-memory.md"))
	for _, carry := range carries {
		info, err := os.Stat(carry)
		if err != nil || info.Size() <= carryLimit {
			continue
		}

		found = append(found, fmt.Sprintf(
			"%s is %d KB, past the %d KB a carry-forward holds. Anything the repository can rebuild goes.",
			filepath.Base(carry), info.Size()/1024, carryLimit/1024,
		))
	}

	index, err := os.ReadFile(filepath.Join(memory, "MEMORY.md"))
	if err != nil {
		return found
	}

	for _, line := range strings.Split(string(index), "\n") {
		for _, match := range memoryLink.FindAllStringSubmatch(line, -1) {
			target := match[1]
			if strings.HasPrefix(target, "http://") || strings.HasPrefix(target, "https://") {
				continue
			}

			_, err := os.Stat(filepath.Join(memory, target))
			if err != nil {
				found = append(found, fmt.Sprintf("MEMORY.md points at %s, which is not there.", target))
			}
		}
	}

	return found
}

// remove reports only whether the directory went. A failure leaves nothing a
// hook's caller could act on, so the next start retries it.
func remove(ctx context.Context, directory string) bool {
	if ctx.Err() != nil {
		return false
	}

	return os.RemoveAll(directory) == nil
}

func isDir(path string) bool {
	info, err := os.Stat(path)

	return err == nil && info.IsDir()
}
